src/coreclr/jit/valuenum.cpp

```diff
@@ -8267,172 +8284,85 @@ void Compiler::fgValueNumberTree(GenTree* tree)
-                                //
-                                ValueNumPair partialLclVarVNP =
-                                    vnStore->VNPairForCast(wholeLclVarVNP, typ, varDsc->TypeGet());
-                                lcl->gtVNPair = partialLclVarVNP;
+                                assert((varDsc->TypeGet() == TYP_LONG) && lcl->TypeIs(TYP_INT));
+                                lcl->gtVNPair =
+                                    vnStore->VNPairForCast(wholeLclVarVNP, lcl->TypeGet(), varDsc->TypeGet());
                             }
                             else
                             {
-                                assert(typSize > varSize);
-                                // the indirection is reading beyond the end of the field
-                                //
-                                generateUniqueVN = true;
+                                assert((varDsc->TypeGet() == TYP_I_IMPL) && lcl->TypeIs(TYP_BYREF));
+                                lcl->gtVNPair = wholeLclVarVNP;
                             }
                         }
+                        else
+                        {
+                            lcl->gtVNPair = wholeLclVarVNP;
+                        }
                     }
-
-                    if (!generateUniqueVN)
+                    else if (varDsc->IsAddressExposed())
                     {
-                        // There are a couple of cases where we haven't assigned a valid value number to 'lcl'
-                        //
-                        if (lcl->gtVNPair.GetLiberal() == ValueNumStore::NoVN)
-                        {
-#ifdef DEBUG
-
-                            // So far, we know about three of these cases:
-                            // Case 1) We have a local var who has never been defined but it's seen as a use.
-                            //         This is the case of storeIndir(addr(lclvar)) = expr.  In this case since we only
-                            //         take the address of the variable, this doesn't mean it's a use nor we have to
-                            //         initialize it, so in this very rare case, we fabricate a value number;
-                            // Case 2) Local variables that represent structs which are assigned using CpBlk;
-                            // Case 3) Local variable was written using a partial write,
-                            //         for example, BLK<1>(ADDR(LCL_VAR int)) = 1, it will change only the first byte.
-                            //         Check that there was ld-addr-op on the local.
-                            //
-                            // Make sure we have one of these cases.
-                            //
-                            const GenTree*   nextNode = lcl->gtNext;
-                            const LclVarDsc* varDsc   = lvaGetDesc(lcl);
-
-                            assert((nextNode->gtOper == GT_ADDR && nextNode->AsOp()->gtOp1 == lcl) ||
```

Comment:
Didn't replicate this assert as I don't find it useful - yes, we will see some locals under `ADDR`s here, but that's just how the IR is.